{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# Dependencies\n",
    "import modules.blast as blast\n",
    "import modules.uniprot as uniprot\n",
    "import modules.msa as msa\n",
    "import pandas as pd\n",
    "import time"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Constants\n",
    "BLAST_MATCHES_PATH = 'data/blast.csv'  # Blast matches dataset\n",
    "BLAST_FASTA_PATH = 'data/blast.fasta'  # Blast matched proteins fasta file\n",
    "MSA_FASTA_PATH = 'data/msa.fasta'  # Multiple sequence alignment file\n",
    "HUMAN_CSV_PATH = 'data/human.csv'  # Test set csv\n",
    "HUMAN_FASTA_PATH = 'data/human.fasta'  # Test set fasta"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## BLAST search\n",
    "\n",
    "Execute a blast search using EBI restful web service"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Started job: ncbiblast-R20200206-223016-0184-83717379-p2m\n"
     ]
    }
   ],
   "source": [
    "# Start a BLAST job\n",
    "status, job_id, _ = blast.run_job(email='damiano.clementel@studenti.unipd.it',\n",
    "                              sequence='VPSGWKAVFDDEYQTWYYVDLSTNSSQWEPP',\n",
    "                              params={'database': 'uniref90',\n",
    "                                      'matrix': 'BLOSUM62',\n",
    "                                      'alignments': 1000, \n",
    "                                      'gapalign': True, \n",
    "                                      'exp': '1e-3'})\n",
    "\n",
    "# Check output\n",
    "print('Started job: {:s}'.format(job_id))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Job status: RUNNING\n",
      "Job status: RUNNING\n",
      "Job status: RUNNING\n",
      "Job status: RUNNING\n",
      "Job status: RUNNING\n",
      "Job status: RUNNING\n",
      "Job status: RUNNING\n",
      "Job status: RUNNING\n",
      "Job status: FINISHED\n",
      "Job exited with status: FINISHED\n"
     ]
    }
   ],
   "source": [
    "# Define results container\n",
    "job_result = None\n",
    "# job_id = 'ncbiblast-R20200206-172705-0918-98731783-p1m'\n",
    "\n",
    "# Retruieve BLAST job results\n",
    "while True:\n",
    "    # Make request for job status\n",
    "    status, job_status, _ = blast.get_job_status(job_id)\n",
    "    print('Job status: {:s}'.format(job_status))  # LOG\n",
    "    # Check error\n",
    "    if not status: break\n",
    "    # Check if job has finished running\n",
    "    if job_status == 'FINISHED':\n",
    "        # Retrieve results\n",
    "        status, job_result, _ = blast.get_job_result(job_id)\n",
    "        print('Job exited with status: {:s}'.format(job_status))  # LOG\n",
    "        break  # Exit loop\n",
    "    # Wait 10 seconds befroe making another call\n",
    "    time.sleep(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>score</th>\n",
       "      <th>bits</th>\n",
       "      <th>expectation</th>\n",
       "      <th>identity</th>\n",
       "      <th>positives</th>\n",
       "      <th>gaps</th>\n",
       "      <th>strand</th>\n",
       "      <th>pattern_seq</th>\n",
       "      <th>match_seq</th>\n",
       "      <th>match_start</th>\n",
       "      <th>match_end</th>\n",
       "      <th>database</th>\n",
       "      <th>id</th>\n",
       "      <th>ac</th>\n",
       "      <th>description</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>182</td>\n",
       "      <td>74.7146</td>\n",
       "      <td>1.000000e-15</td>\n",
       "      <td>100.0</td>\n",
       "      <td>100.0</td>\n",
       "      <td>0</td>\n",
       "      <td>none/none</td>\n",
       "      <td>VPSGWKAVFDDEYQTWYYVDLSTNSSQWEPP</td>\n",
       "      <td>VPSGWKAVFDDEYQTWYYVDLSTNSSQWEPP</td>\n",
       "      <td>11</td>\n",
       "      <td>41</td>\n",
       "      <td>UR90</td>\n",
       "      <td>UniRef90_P43582</td>\n",
       "      <td>WW</td>\n",
       "      <td>WW domain-containing protein WWM1 n=9 Tax=Sacc...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>175</td>\n",
       "      <td>72.0182</td>\n",
       "      <td>1.100000e-14</td>\n",
       "      <td>93.5</td>\n",
       "      <td>100.0</td>\n",
       "      <td>0</td>\n",
       "      <td>none/none</td>\n",
       "      <td>VPSGWKAVFDDEYQTW+YVDLSTN+SQWEPP</td>\n",
       "      <td>VPSGWKAVFDDEYQTWFYVDLSTNNSQWEPP</td>\n",
       "      <td>11</td>\n",
       "      <td>41</td>\n",
       "      <td>UR90</td>\n",
       "      <td>UniRef90_J8Q8J2</td>\n",
       "      <td>Wwm1p</td>\n",
       "      <td>Wwm1p n=1 Tax=Saccharomyces arboricola (strain...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>173</td>\n",
       "      <td>71.2478</td>\n",
       "      <td>1.800000e-14</td>\n",
       "      <td>93.5</td>\n",
       "      <td>100.0</td>\n",
       "      <td>0</td>\n",
       "      <td>none/none</td>\n",
       "      <td>VPSGWKAVFDDEYQTW+YV+LSTNSSQWEPP</td>\n",
       "      <td>VPSGWKAVFDDEYQTWFYVNLSTNSSQWEPP</td>\n",
       "      <td>11</td>\n",
       "      <td>41</td>\n",
       "      <td>UR90</td>\n",
       "      <td>UniRef90_J5RH20</td>\n",
       "      <td>WWM1-like</td>\n",
       "      <td>WWM1-like protein n=2 Tax=Saccharomyces TaxID=...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>170</td>\n",
       "      <td>70.0922</td>\n",
       "      <td>4.700000e-14</td>\n",
       "      <td>93.5</td>\n",
       "      <td>96.8</td>\n",
       "      <td>0</td>\n",
       "      <td>none/none</td>\n",
       "      <td>VPSGWKAVFDDEYQTW+YVDLSTNSSQWE P</td>\n",
       "      <td>VPSGWKAVFDDEYQTWFYVDLSTNSSQWEAP</td>\n",
       "      <td>11</td>\n",
       "      <td>41</td>\n",
       "      <td>UR90</td>\n",
       "      <td>UniRef90_A0A0L8RJY2</td>\n",
       "      <td>WWM1-like</td>\n",
       "      <td>WWM1-like protein n=1 Tax=Saccharomyces eubaya...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>158</td>\n",
       "      <td>65.4698</td>\n",
       "      <td>2.800000e-12</td>\n",
       "      <td>80.6</td>\n",
       "      <td>93.5</td>\n",
       "      <td>0</td>\n",
       "      <td>none/none</td>\n",
       "      <td>VP GWKAVFDDEY+TW+YV+L+TN SQWEPP</td>\n",
       "      <td>VPKGWKAVFDDEYKTWFYVNLATNQSQWEPP</td>\n",
       "      <td>11</td>\n",
       "      <td>41</td>\n",
       "      <td>UR90</td>\n",
       "      <td>UniRef90_A0A212M9M4</td>\n",
       "      <td>WW</td>\n",
       "      <td>WW domain-containing protein n=4 Tax=Zygosacch...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   score     bits   expectation  identity  positives  gaps     strand  \\\n",
       "0    182  74.7146  1.000000e-15     100.0      100.0     0  none/none   \n",
       "1    175  72.0182  1.100000e-14      93.5      100.0     0  none/none   \n",
       "2    173  71.2478  1.800000e-14      93.5      100.0     0  none/none   \n",
       "3    170  70.0922  4.700000e-14      93.5       96.8     0  none/none   \n",
       "4    158  65.4698  2.800000e-12      80.6       93.5     0  none/none   \n",
       "\n",
       "                       pattern_seq                        match_seq  \\\n",
       "0  VPSGWKAVFDDEYQTWYYVDLSTNSSQWEPP  VPSGWKAVFDDEYQTWYYVDLSTNSSQWEPP   \n",
       "1  VPSGWKAVFDDEYQTW+YVDLSTN+SQWEPP  VPSGWKAVFDDEYQTWFYVDLSTNNSQWEPP   \n",
       "2  VPSGWKAVFDDEYQTW+YV+LSTNSSQWEPP  VPSGWKAVFDDEYQTWFYVNLSTNSSQWEPP   \n",
       "3  VPSGWKAVFDDEYQTW+YVDLSTNSSQWE P  VPSGWKAVFDDEYQTWFYVDLSTNSSQWEAP   \n",
       "4  VP GWKAVFDDEY+TW+YV+L+TN SQWEPP  VPKGWKAVFDDEYKTWFYVNLATNQSQWEPP   \n",
       "\n",
       "   match_start  match_end database                   id         ac  \\\n",
       "0           11         41     UR90      UniRef90_P43582         WW   \n",
       "1           11         41     UR90      UniRef90_J8Q8J2      Wwm1p   \n",
       "2           11         41     UR90      UniRef90_J5RH20  WWM1-like   \n",
       "3           11         41     UR90  UniRef90_A0A0L8RJY2  WWM1-like   \n",
       "4           11         41     UR90  UniRef90_A0A212M9M4         WW   \n",
       "\n",
       "                                         description  \n",
       "0  WW domain-containing protein WWM1 n=9 Tax=Sacc...  \n",
       "1  Wwm1p n=1 Tax=Saccharomyces arboricola (strain...  \n",
       "2  WWM1-like protein n=2 Tax=Saccharomyces TaxID=...  \n",
       "3  WWM1-like protein n=1 Tax=Saccharomyces eubaya...  \n",
       "4  WW domain-containing protein n=4 Tax=Zygosacch...  "
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Turn job result into pandas DataFrame object\n",
    "matches = pd.DataFrame(job_result)\n",
    "matches.to_csv(BLAST_MATCHES_PATH, index=False)  # Save to disk\n",
    "\n",
    "# Show dataframe\n",
    "matches.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Create FASTA file from BLAST"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save full fasta file\n",
    "with open(BLAST_FASTA_PATH, 'w') as blast_fasta_file:\n",
    "    # Get fasta file\n",
    "    status, blast_fasta, _ = uniprot.get_proteins(matches.id.tolist(), batch_size=100)\n",
    "    # Write fasta file to disk\n",
    "    blast_fasta_file.write(blast_fasta)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Check number of retrieved fasta files\n",
    "blast_fasta_len = sum([1 for c in blast_fasta if c == '>'])  # Number of rows in fasta file\n",
    "matches_len = matches.shape[0]  # Number of rows in dataframe\n",
    "\n",
    "assert  blast_fasta_len == matches_len, 'Fasta file and dataset lengths do not coincide'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Execute multiple sequence alignment"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Job status: RUNNING\n",
      "Job status: RUNNING\n",
      "Job status: RUNNING\n",
      "Job status: RUNNING\n",
      "Job status: RUNNING\n",
      "Job status: RUNNING\n",
      "Job status: RUNNING\n",
      "Job status: RUNNING\n",
      "Job status: RUNNING\n",
      "Job status: RUNNING\n",
      "Job status: RUNNING\n",
      "Job status: RUNNING\n",
      "Job status: RUNNING\n",
      "Job status: RUNNING\n",
      "Job status: RUNNING\n",
      "Job status: FINISHED\n",
      "\n",
      "Retrieved fasta file\n",
      ">UniRef90_P43582 WW domain-containing protein WWM1 n=9 Tax=Saccharomyces TaxID=4930 RepID=WWM1_YEAST\n",
      "------------------------------------------------------------\n",
      "------------------------------------------------------------\n",
      "------------------------------------------------------------\n",
      "------------------------------------------------------------\n",
      "------------------------------------------------------------\n",
      "------------------------------------------------------------\n",
      "------------------------------------------------------------\n",
      "------------------------------------------------------------\n",
      "------------------------------------------------------------\n",
      "------------------------------------------------------------\n",
      "------------------------------------------------------------\n",
      "------------------------------------------------------------\n",
      "------------------------------------------------------------\n",
      "------------------------------------------------------------\n",
      "------------------------------------------------------------\n",
      "------------------------------------------------------------\n",
      "------------------------------------------------------------\n",
      "------------------------------------------------------------\n",
      "------------------------------------------------------------\n",
      "------------------------------------------------------------\n",
      "------------------------------------------------------------\n",
      "------------------------------------------------------------\n",
      "------------------------------------------------------------\n",
      "------------------------------------------------------------\n",
      "------------------------------------------------------------\n",
      "------------------------------------------------------------\n",
      "------------------------------------------------------------\n",
      "------------------------------------------------------------\n",
      "------------------------------------------------------------\n",
      "------------------------------------------------------------\n",
      "------------------------------------------------------------\n",
      "--------\n"
     ]
    }
   ],
   "source": [
    "# Make multiple sequence alignment\n",
    "\n",
    "# Create MSA job\n",
    "status, job_id, _ = msa.run_clustalo(email='damiano.clementel@studenti.unipd.it', sequence=blast_fasta)\n",
    "\n",
    "# Wait until MSA job finishes\n",
    "while True:\n",
    "    # Check response status\n",
    "    status, job_status, _ = msa.get_job_status(job_id, algorithm=msa.CLUSTALO)\n",
    "    # Check if response status is 200 OK\n",
    "    if not status:\n",
    "        print('Error: job exited')\n",
    "        break \n",
    "    # Check if job has finished, then go on\n",
    "    print('Job status: {:s}'.format(job_status))\n",
    "    if job_status == 'FINISHED': break\n",
    "    # Add delay\n",
    "    time.sleep(3)\n",
    "print()\n",
    "    \n",
    "# Get results\n",
    "if status:\n",
    "    # Retrieve result as fasta\n",
    "    status, msa_fasta, _ = msa.get_job_result(job_id, algorithm=msa.CLUSTALO)\n",
    "    # Save msa to file\n",
    "    with open(MSA_FASTA_PATH, 'w') as msa_fasta_file:\n",
    "        msa_fasta_file.write(msa_fasta)\n",
    "    # Show fasta file head\n",
    "    print('Retrieved fasta file')\n",
    "    print(msa_fasta[:2000])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Check length of retrieved fasta\n",
    "msa_fasta_len = sum([1 for c in blast_fasta if c == '>'])  # Number of rows in fasta file\n",
    "\n",
    "assert blast_fasta_len == matches_len, 'Alignments fasta file and dataset lengths do not coincide'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Create a dataset of human proteins"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Retrieve dataset of human proteins\n",
    "status, table, _ = uniprot.make_query('organism:\"Homo sapiens (Human) [9606]\"', params={\n",
    "    'compress': 'no',\n",
    "    'columns': ','.join(['id', 'entry name', 'reviewed', 'protein names', 'genes', 'length']),\n",
    "    'sort': 'score',\n",
    "    'format': 'tab'\n",
    "})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Entry</th>\n",
       "      <th>Entry name</th>\n",
       "      <th>Status</th>\n",
       "      <th>Protein names</th>\n",
       "      <th>Gene names</th>\n",
       "      <th>Length</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Q9Y263</td>\n",
       "      <td>PLAP_HUMAN</td>\n",
       "      <td>reviewed</td>\n",
       "      <td>Phospholipase A-2-activating protein (PLA2P) (...</td>\n",
       "      <td>PLAA PLAP</td>\n",
       "      <td>795</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Q96RE7</td>\n",
       "      <td>NACC1_HUMAN</td>\n",
       "      <td>reviewed</td>\n",
       "      <td>Nucleus accumbens-associated protein 1 (NAC-1)...</td>\n",
       "      <td>NACC1 BTBD14B NAC1</td>\n",
       "      <td>527</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>O43312</td>\n",
       "      <td>MTSS1_HUMAN</td>\n",
       "      <td>reviewed</td>\n",
       "      <td>Protein MTSS 1 (Metastasis suppressor YGL-1) (...</td>\n",
       "      <td>MTSS1 KIAA0429 MIM</td>\n",
       "      <td>755</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Q9NP80</td>\n",
       "      <td>PLPL8_HUMAN</td>\n",
       "      <td>reviewed</td>\n",
       "      <td>Calcium-independent phospholipase A2-gamma (EC...</td>\n",
       "      <td>PNPLA8 IPLA22 IPLA2G BM-043</td>\n",
       "      <td>782</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Q15319</td>\n",
       "      <td>PO4F3_HUMAN</td>\n",
       "      <td>reviewed</td>\n",
       "      <td>POU domain, class 4, transcription factor 3 (B...</td>\n",
       "      <td>POU4F3 BRN3C</td>\n",
       "      <td>338</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    Entry   Entry name    Status  \\\n",
       "0  Q9Y263   PLAP_HUMAN  reviewed   \n",
       "1  Q96RE7  NACC1_HUMAN  reviewed   \n",
       "2  O43312  MTSS1_HUMAN  reviewed   \n",
       "3  Q9NP80  PLPL8_HUMAN  reviewed   \n",
       "4  Q15319  PO4F3_HUMAN  reviewed   \n",
       "\n",
       "                                       Protein names  \\\n",
       "0  Phospholipase A-2-activating protein (PLA2P) (...   \n",
       "1  Nucleus accumbens-associated protein 1 (NAC-1)...   \n",
       "2  Protein MTSS 1 (Metastasis suppressor YGL-1) (...   \n",
       "3  Calcium-independent phospholipase A2-gamma (EC...   \n",
       "4  POU domain, class 4, transcription factor 3 (B...   \n",
       "\n",
       "                    Gene names Length  \n",
       "0                    PLAA PLAP    795  \n",
       "1           NACC1 BTBD14B NAC1    527  \n",
       "2           MTSS1 KIAA0429 MIM    755  \n",
       "3  PNPLA8 IPLA22 IPLA2G BM-043    782  \n",
       "4                 POU4F3 BRN3C    338  "
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Create DataFrame object\n",
    "\n",
    "# Get rows and header row\n",
    "rows = table.split('\\n')\n",
    "header = rows[0].split('\\t')\n",
    "\n",
    "# Instantiate new dataframe\n",
    "human_ds = pd.DataFrame([row.split('\\t') for row in rows[1:]], columns=header)\n",
    "human_ds.to_csv(HUMAN_CSV_PATH, index=False)\n",
    "human_ds.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Entry</th>\n",
       "      <th>Entry name</th>\n",
       "      <th>Status</th>\n",
       "      <th>Protein names</th>\n",
       "      <th>Gene names</th>\n",
       "      <th>Length</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>H0YNE8</td>\n",
       "      <td>H0YNE8_HUMAN</td>\n",
       "      <td>unreviewed</td>\n",
       "      <td>Amyloid-beta A4 precursor protein-binding fami...</td>\n",
       "      <td>APBA2</td>\n",
       "      <td>27</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>F8W0M7</td>\n",
       "      <td>F8W0M7_HUMAN</td>\n",
       "      <td>unreviewed</td>\n",
       "      <td>Methionyl-tRNA synthetase 1</td>\n",
       "      <td>MARS1</td>\n",
       "      <td>95</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>G3V1D8</td>\n",
       "      <td>G3V1D8_HUMAN</td>\n",
       "      <td>unreviewed</td>\n",
       "      <td>Leucine rich repeat containing 24 (Leucine-ric...</td>\n",
       "      <td>LRRC24 hCG_1818221</td>\n",
       "      <td>510</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>B2XJG5</td>\n",
       "      <td>B2XJG5_HUMAN</td>\n",
       "      <td>unreviewed</td>\n",
       "      <td>NADH-ubiquinone oxidoreductase chain 4 (EC 7.1...</td>\n",
       "      <td>ND4</td>\n",
       "      <td>459</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>B2XHT9</td>\n",
       "      <td>B2XHT9_HUMAN</td>\n",
       "      <td>unreviewed</td>\n",
       "      <td>Cytochrome c oxidase subunit 2</td>\n",
       "      <td>COX2</td>\n",
       "      <td>227</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     Entry    Entry name      Status  \\\n",
       "26  H0YNE8  H0YNE8_HUMAN  unreviewed   \n",
       "27  F8W0M7  F8W0M7_HUMAN  unreviewed   \n",
       "28  G3V1D8  G3V1D8_HUMAN  unreviewed   \n",
       "29  B2XJG5  B2XJG5_HUMAN  unreviewed   \n",
       "30  B2XHT9  B2XHT9_HUMAN  unreviewed   \n",
       "\n",
       "                                        Protein names          Gene names  \\\n",
       "26  Amyloid-beta A4 precursor protein-binding fami...               APBA2   \n",
       "27                        Methionyl-tRNA synthetase 1               MARS1   \n",
       "28  Leucine rich repeat containing 24 (Leucine-ric...  LRRC24 hCG_1818221   \n",
       "29  NADH-ubiquinone oxidoreductase chain 4 (EC 7.1...                 ND4   \n",
       "30                     Cytochrome c oxidase subunit 2                COX2   \n",
       "\n",
       "   Length  \n",
       "26     27  \n",
       "27     95  \n",
       "28    510  \n",
       "29    459  \n",
       "30    227  "
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Get non reviewed proteins\n",
    "human_ds[human_ds.Status != 'reviewed'].head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Entry</th>\n",
       "      <th>Entry name</th>\n",
       "      <th>Status</th>\n",
       "      <th>Protein names</th>\n",
       "      <th>Gene names</th>\n",
       "      <th>Length</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Q9Y263</td>\n",
       "      <td>PLAP_HUMAN</td>\n",
       "      <td>reviewed</td>\n",
       "      <td>Phospholipase A-2-activating protein (PLA2P) (...</td>\n",
       "      <td>PLAA PLAP</td>\n",
       "      <td>795</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Q96RE7</td>\n",
       "      <td>NACC1_HUMAN</td>\n",
       "      <td>reviewed</td>\n",
       "      <td>Nucleus accumbens-associated protein 1 (NAC-1)...</td>\n",
       "      <td>NACC1 BTBD14B NAC1</td>\n",
       "      <td>527</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>O43312</td>\n",
       "      <td>MTSS1_HUMAN</td>\n",
       "      <td>reviewed</td>\n",
       "      <td>Protein MTSS 1 (Metastasis suppressor YGL-1) (...</td>\n",
       "      <td>MTSS1 KIAA0429 MIM</td>\n",
       "      <td>755</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Q9NP80</td>\n",
       "      <td>PLPL8_HUMAN</td>\n",
       "      <td>reviewed</td>\n",
       "      <td>Calcium-independent phospholipase A2-gamma (EC...</td>\n",
       "      <td>PNPLA8 IPLA22 IPLA2G BM-043</td>\n",
       "      <td>782</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Q15319</td>\n",
       "      <td>PO4F3_HUMAN</td>\n",
       "      <td>reviewed</td>\n",
       "      <td>POU domain, class 4, transcription factor 3 (B...</td>\n",
       "      <td>POU4F3 BRN3C</td>\n",
       "      <td>338</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    Entry   Entry name    Status  \\\n",
       "0  Q9Y263   PLAP_HUMAN  reviewed   \n",
       "1  Q96RE7  NACC1_HUMAN  reviewed   \n",
       "2  O43312  MTSS1_HUMAN  reviewed   \n",
       "3  Q9NP80  PLPL8_HUMAN  reviewed   \n",
       "4  Q15319  PO4F3_HUMAN  reviewed   \n",
       "\n",
       "                                       Protein names  \\\n",
       "0  Phospholipase A-2-activating protein (PLA2P) (...   \n",
       "1  Nucleus accumbens-associated protein 1 (NAC-1)...   \n",
       "2  Protein MTSS 1 (Metastasis suppressor YGL-1) (...   \n",
       "3  Calcium-independent phospholipase A2-gamma (EC...   \n",
       "4  POU domain, class 4, transcription factor 3 (B...   \n",
       "\n",
       "                    Gene names Length  \n",
       "0                    PLAA PLAP    795  \n",
       "1           NACC1 BTBD14B NAC1    527  \n",
       "2           MTSS1 KIAA0429 MIM    755  \n",
       "3  PNPLA8 IPLA22 IPLA2G BM-043    782  \n",
       "4                 POU4F3 BRN3C    338  "
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Get reviewed proteins\n",
    "human_ds[human_ds.Status == 'reviewed'].head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Download fasta file\n",
    "status, human_fasta, _ =  uniprot.make_query('organism:\"Homo sapiens (Human) [9606]\"', params={\n",
    "    'compress': 'no',\n",
    "    'sort': 'score',\n",
    "    'format': 'fasta'\n",
    "})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      ">sp|Q9Y263|PLAP_HUMAN Phospholipase A-2-activating protein OS=Homo sapiens OX=9606 GN=PLAA PE=1 SV=2\n",
      "MTSGATRYRLSCSLRGHELDVRGLVCCAYPPGAFVSVSRDRTTRLWAPDSPNRSFTEMHC\n",
      "MSGHSNFVSCVCIIPSSDIYPHGLIATGGNDHNICIFSLDSPMPLYILKGHKNTVCSLSS\n",
      "GKFGTLLSGSWDTTAKVWLNDKCMMTLQGHTAAVWAVKILPEQGLMLTGSADKTVKLWKA\n",
      "GRCERTFSGHEDCVRGLAILSETEFLSCANDASIRRWQITGECLEVYYGHTNYIYSISVF\n",
      "PNCRDFVTTAEDRSLRIWKHGECAQTIRLPAQSIWCCCVLDNGDIVVGASDGIIRVFTES\n",
      "EDRTASAEEIKAFEKELSHATIDSKTGDLGDINAEQLPGREHLNEPGTREGQTRLIRDGE\n",
      "KVEAYQWSVSEGRWIKIGDVVGSSGANQQTSGKVLYEGKEFDYVFSIDVNEGGPSYKLPY\n",
      "NTSDDPWLTAYNFLQKNDLNPMFLDQVAKFIIDNTKGQMLGLGNPSFSDPFTGGGRYVPG\n",
      "SSGSSNTLPTADPFTGAGRYVPGSASMGTTMAGVDPFTGNSAYRSAASKTMNIYFPKKEA\n",
      "VTFDQANPTQILGKLKELNGTAPEEKKLTEDDLILLEKILSLICNSSSEKPTVQQLQILW\n",
      "KAINCPEDIVFPALDILRLSIKHPSVNENFCNEKEGAQFSSHLINLLNPKGKPANQLLAL\n",
      "RTFCNCFVGQAGQKLMMSQRESLMSHAIELKSGSNKNIHIALATLALNYSVCFHKDHNIE\n",
      "GKAQCLSLISTILEVVQDLEATFRLLVALGTLISDDSNAVQLAKSLGVDSQIKKYSSVSE\n",
      "PAKVSECCRFILNLL\n",
      ">sp|Q96RE7|NACC1_HUMAN Nucleus accumbens-associated protein 1 OS=Homo sapiens OX=9606 GN=N\n"
     ]
    }
   ],
   "source": [
    "# Check the retrieved fasta file\n",
    "print(human_fasta[:1000])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save human fasta to disk\n",
    "with open(HUMAN_FASTA_PATH, 'w') as human_fasta_file:\n",
    "    human_fasta_file.write(human_fasta)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
